src/annotation-engine/use-annotation-engine.ts

```diff
@@ -196,21 +263,43 @@ const useAnnotationEngine = ({
     // Initialize canvas
     useEffect(() => {
         const currentCanvasRef = canvasRef.current;
+        styledAnnotations.current = annotationsToStyle;
 
```

Comment:
ici j'injecte les `annotationsToStyle` gérées dans mon userland, pour surcharger les styles gérés en interne par l'engine

Reply:
il faudrait que tu m'expliques la logique :)

Reply:
L'engine stocke les styles dans la Map `styledAnnotations`. A chaque fois qu'on utilise l'objet `operations` pour styliser une forme (par ex. survol de la souris sur une forme), ça aboutit en définitive à une modification de cette Map.

Je veux pouvoir modifier les styles sans passer par l'objet `operations` : par ex. en cliquant sur le nom d'une roaddata dans la liste déroulante. Du coup, côté userland, je génère une Map similaire à `styledAnnotations`, nommée `annotationsToSyle`, que j'envoie à l'engine en tant que prop, au même titre que `annotations`, `annotationToEdit`...

J'ai modifié le `useEffect` d'initialisation du canvas pour qu'à chaque fois que `annotationsToSyle` change, il vienne écraser les styles de l'engine.

D'après les tests que j'ai faits ça marche très bien.

Il y a quand même un truc qui me chiffonne. Le fait de pouvoir modifier le style soit depuis l'engine (avec `operations`) soit depuis l'extérieur (via la props passée à l'engine), en court-circuitant les `operations`. La responsabilité de styliser est gérée par les deux en même temps, ça ne va pas... 

Plus j'y pense et plus je me dis qu'il va falloir que je retire des `operations` tout ce qui touche à la stylisation... qu'en penses-tu ?

Reply:
(J'ai répondu à tes commentaires en partant du bas puis en remontant jusqu'ici ici 😅 ) Donc ouais c'est un sujet que tu abordes plus bas. Lis ce que j'ai écrit plus bas pour avoir mon avis sur la question.

tldr;
c'est chaud quand même, demande à Manu 😝 

Cela dit j'ai une autre idée, dans le `useAnnotationEngine`, on utilise le hook `useImperativeHandle` qui permet d'offrir au UL la méthode `cancelCreation`. Ce qui permet ensuite de faire un truc du genre:

```typescript
// dans le user-land
const refAE = useRef<Handles>(null);
//...
refAE.cancelCreation();

return <AnnotationEngine
            ref={refAE}
            //...
        />
```

Et donc, why not avoir toutes les méthodes de `operations` dedans, et ne plus refiler `operations` dans chaque event. ??

Reply:
justement, j'avais pas très bien saisi ce système de refAE.
Si je comprends bien, ça permet de "faire remonter" jusqu'à userland des méthodes définies dans le coeur de l'engine ?

Reply:
la bonne approche avec react est de ne pas avoir d'état, être en pure fonctionnel (input » output)
le faite d'avoir le `operations` en paramètre des events est dans cet esprit. Pour chaque event on a un ensemble d'actions possible. Aucune garantie que ce soit la même instance d'operation pour deux events. Si on a plusieurs Annoations engine, pas de soucis…
`useImperativeHandle` est fondamentalement un brèche dans cette approche. Il est nécessaire du faite que l'élément html canvas contient un état, donc on peut toujours avoir une approche fonctionnelle.
donc utiliser le `useImperativeHandle` à chaque fois qu'on a besoin d'envoyer n'importe quand des commandes au canvas: ok on a pas le choix.
Mais attention de ne pas succomber au tout imperatif: dès que c'est possible il faut rester sur une approche impérative, donc laisser le `operation` en l'état et l'enrichir si certaines actions ne sont utilisé que suite à un event.

Reply:
Là le soucis c'est que Operation contient des méthodes que Laurent veut utiliser en dehors des events.

Reply:
Du coup, ma suggestion était d'extraire d'Operation toutes les méthodes qui relèvent de la stylisation, et les appeler exclusivement depuis Userland.
Cela éviterait de recourir au `useImperativeHandle`.
En revanche, ce que j'ai du mal à mesurer, c'est en quoi il est dommageable pour l'AE de ne plus posséder ces méthodes de stylisation.

Reply:
du coup j'ai extrait ces méthodes.
Done